controllers/volumereplicationgroup_controller.go

```diff
@@ -199,38 +203,147 @@ func (v *VolumeReplicationGroupReconciler) Reconcile(ctx context.Context, req ct
+	if volRepGroup.GetDeletionTimestamp().IsZero() {
+		if err := v.AddFinalizer(ctx, volRepGroup, vrgFinalizer); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to add the finalizer to VRG CR %s (%w)", req.NamespacedName, err)
+		}
+	} else {
+		if err := v.RemoveFinalizer(ctx, volRepGroup, vrgFinalizer); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to remove the finalizer from VRG CR %s (%w)", req.NamespacedName, err)
+		}
+
+		// stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
+
+	pvcList, err := v.ListPersistentVolumeClaims(ctx, volRepGroup)
 	if err != nil {
 		log.Error(err, "Handling of Persistent Volume Claims of application failed")
 
 		return ctrl.Result{}, err
 	}
 
 	requeue := false
+	requeue = v.HandlePersistentVolumeClaims(ctx, volRepGroup, pvcList)
 
-	for _, pvc := range pvcList.Items {
-		volRep := &volrep.VolumeReplication{}
+	if requeue {
+		return ctrl.Result{Requeue: requeue}, fmt.Errorf("failed to create VolumeReplication CRs %w", err)
+	}
 
-		err = v.Get(ctx, types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}, volRep)
-		if err != nil {
-			if errors.IsNotFound(err) {
-				if err = v.createVolumeReplicationCRForPVC(ctx, pvc); err == nil {
-					continue
-				}
-			}
+	return ctrl.Result{}, nil
+}
 
-			// requeue on failure to ensure any PVC not having a corresponding VR CR
-			requeue = true
+// It might be better move the helper functions like these to a separate
+// package or a separate go file?
+func containsString(slice []string, s string) bool {
+	for _, item := range slice {
+		if item == s {
+			return true
+		}
+	}
+
+	return false
+}
 
-			log.Error(err, "Failed to get or create VolumeReplication CR for PVC", pvc.Name, pvc.Namespace)
+func removeString(slice []string, s string) (result []string) {
+	for _, item := range slice {
+		if item == s {
+			continue
 		}
+
+		result = append(result, item)
 	}
 
-	return ctrl.Result{Requeue: requeue}, nil
+	return
 }
 
-// HandlePersistentVolumeClaims handles PVCs in the VRG
-func (v *VolumeReplicationGroupReconciler) HandlePersistentVolumeClaims(
+func (v *VolumeReplicationGroupReconciler) AddFinalizer(
+	ctx context.Context,
+	volRepGroup *ramendrv1alpha1.VolumeReplicationGroup,
+	finalizer string) error {
+	if !containsString(volRepGroup.ObjectMeta.Finalizers, finalizer) {
+		volRepGroup.ObjectMeta.Finalizers = append(volRepGroup.ObjectMeta.Finalizers, finalizer)
+		if err := v.Update(context.Background(), volRepGroup); err != nil {
+			return fmt.Errorf("failed to add the finalizer to VRG CR %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (v *VolumeReplicationGroupReconciler) RemoveFinalizer(
+	ctx context.Context,
+	volRepGroup *ramendrv1alpha1.VolumeReplicationGroup,
+	finalizer string) error {
+	if containsString(volRepGroup.ObjectMeta.Finalizers, finalizer) {
+		if err := v.deleteRelatedItems(ctx, volRepGroup); err != nil {
```

Comment:
I suggest moving this line outside RemoveFinalizer().  In other words, let RemoveFinalizer only removes the finalizer from the K8s object.  A possible change would be made in line 212 and would be something something like:
```
         if containsString(volRepGroup.ObjectMeta.Finalizers, finalizer) {
		if err := v.deleteRelatedItems(ctx, volRepGroup); err != nil {
		        ........
		}
        }

        if err := v.RemoveFinalizer(ctx, volRepGroup, vrgFinalizer); err != nil {
			return ctrl.Result{}, fmt.Errorf("failed to remove the finalizer from VRG CR %s (%w)", req.NamespacedName, err)
		}
```

Reply:
> I suggest moving this line outside RemoveFinalizer(). In other words, let RemoveFinalizer only removes the finalizer from the K8s object. A possible change would be made in line 212 and would be something something like:
> 
> ```
>          if containsString(volRepGroup.ObjectMeta.Finalizers, finalizer) {
> 		if err := v.deleteRelatedItems(ctx, volRepGroup); err != nil {
> 		        ........
> 		}
>         }
> 

So, line 212 is from Reconcile function and its length would exceed 60 lines failing the linter check. Alternatively what I can do is this:

- Let RemoveFinalizer only remove the finalizer.
- Add another function (say CheckAndRemoveFinalier) which would do the checking of whether the finalizer is
  there or not and if present, then call RemoveFinalizer.
- From reconcile (i.e. line 212) call "CheckAndRemoveFinalizer"

Does it look ok?


>         if err := v.RemoveFinalizer(ctx, volRepGroup, vrgFinalizer); err != nil {
> 			return ctrl.Result{}, fmt.Errorf("failed to remove the finalizer from VRG CR %s (%w)", req.NamespacedName, err)
> 		}
> ```

Reply:
That works

Reply:
ACK